### DISCUSSION TITLE
Error with MIfitHMM

### DISCUSSION THREAD

**Discussion by sarahchinn** [bmcclintock/momentuHMM#138] [Q&A]
I am attempting to follow Theo Michelot's tutorial from his seminar with the Ecological Forecasting Initiative on 2022 Feb 7. I am working with GPS collar data (only GPS, no Argos) for polar bears and I need to regularize the track to evenly spaced time intervals (60 min) and am doing so with the crawlWrap function. I played around with the initial parameters and think I have found some that make sense. The model output looks believable for a 2-state single track (but maybe I'm wrong, see below for workflow and model output):
### Load the data
raw <- read.csv("PB21615.csv") #attached
[PB21615.csv](https://github.com/bmcclintock/momentuHMM/files/8544367/PB21615.csv)

head(raw)

### keep relevant columns: animal, datetime, longitude, latitude, x, y (LAEA projection)
data_all <- raw[,c(1,4,7,8,22, 23)]
head(data_all)
colnames(data_all) <- c("ID", "time", "lat", "lon", "x", "y")
data_all$time <- as.POSIXct(data_all$time, format= "%m/%d/%Y %H:%M", tz="GMT")

data <- data_all

### Remove duplicate datetimes
dups <- which(duplicated(data$time)) #none in this bear's data

### Plot Northing vs Easting (in LAEA centered on Wrangel Island, Russia)
ggplot(data, aes(x, y, col = ID)) +
  geom_point(size = 0.5) + geom_path() +
  coord_equal()

### Table of time intervals in data - regularize for one location every 60 minutes
plot(table(diff(data$time)), xlim = c(0, 60), 
     xlab = "time interval (min)", ylab = "count")

### Use function from utility_function.R to split data at gaps > 14 days
data_split <- split_at_gap(data = data, max_gap = 336*60, shortest_track = 24*60)

### Predict locations on 60-min grid using crawl (through momentuHMM wrapper)
ln.prior <- function(theta) dnorm(theta[2], -4,2,log=T)

crw_out <- crawlWrap(obsData = data_split, 
                     timeStep = "60 min",
                     Time.name = "time", 
                     coord = c("x", "y"), 
                     theta = c(log(10),3), 
                     fixPar=c(NA,NA), 
                     prior=ln.prior, 
                     initialSANN = list(maxit = 5000), 
                     retryFits=10, 
                     attempts=10)

crw_out1 <- crawlWrap(obsData = data_split, timeStep = "60 min",
                     Time.name = "time", coord = c("lon", "lat"), initialSANN = list(maxit = 5000), retryFits=10)

### The output can then directly be passed to prepData to get step lengths and turning angles.
data_hmm <- prepData(data = crw_out) 

### data preparation
data_hmm <- prepData(data = crw_out) 

### Plot histogram of step lengths and turn angles
hist(data_hmm$step, xlim=c(0,300), breaks=100) #most step lengths are 20m or less
hist(data_hmm$angle, breaks=seq(-pi, pi, length=15))

### Observation distribution (step lengths and turning angles)
dist <- list(step = "gamma", angle = "vm")

Par0_2s <- list(step = c(10, 100, 50, 30), angle = c(.03, 2)) 

### Fit a 2-state HMM
states <- c("ARS/rest", "Traveling")
hmm1 <- fitHMM(data_hmm, nbStates = 2, dist = dist, Par0 = Par0_2s, stateNames=states)

### print parameter estimates
hmm1

### Plot estimated distributions and state-coloured tracks
plot(hmm1, breaks = 10, ask = T)

###############################################################
Then I attempt to use the multiple imputation function to account for the uncertainty in predicted locations. But, when I run the function MIfitHMM I get a convergence error (Hessian is singular - I have copied the errors below the code). I tried to fix this by adding additional arguments to the crawlWrap function (retryFits, attempts, and added theta, prior, fixPar) but I still get the same errors. In this example I am only running nSims=2 to decrease processing time, but even when I run with more simulations I still get the same errors (more NaNs produced). I'm am guessing I need to adjust something within the crawlWrap function but have yet to be successful. Thanks for the help!

## Fit HMM using multiple imputation
library(mitools)
library(doFuture)
hmm6 <- MIfitHMM(miData = crw_out, nSims = 2, ncores = 1, nbStates = 2, 
                 dist = dist, Par0 = Par0_2s, retryFits = 50)
hmm6

### Errors ####
Warning messages:
1: In MIpool(fits, alpha = alpha, ncores = ncores, na.rm = na.rm) :
  Hessian is singular for HMM fit(s): 1, 2
2: In sqrt(diag(x)) : NaNs produced
3: In MIpool(fits, alpha = alpha, ncores = ncores, na.rm = na.rm) :
  working parameter standard errors are not finite for HMM fits 2
4: In sqrt(diag(x)) : NaNs produced

**Comment by bmcclintock** [comment]
Those are not errors, they are warnings related to the estimation of the standard errors of the HMM parameters based on the Hessian matrix. Without the `split_at_gap` function, I'm unable to reproduce your example. But if I had to guess, you likely have parameters being estimated at the (real scale) boundary when fitting the HMMs with `MIfitHMM`. Have you examined the model output?
